core/trino-main/src/main/java/io/trino/sql/planner/LogicalPlanner.java

```diff
@@ -451,29 +453,39 @@ private RelationPlan getInsertPlan(
                 continue;
             }
             Symbol output = symbolAllocator.newSymbol(column.getName(), column.getType());
+            Expression expression;
+            Type tableType = column.getType();
             int index = insertColumns.indexOf(columns.get(column.getName()));
             if (index < 0) {
                 if (supportsMissingColumnsOnInsert) {
                     continue;
                 }
-                Expression cast = new Cast(new NullLiteral(), toSqlType(column.getType()));
-                assignments.put(output, cast);
-                insertedColumnsBuilder.add(column);
+                if (!column.isNullable()) {
+                    expression = createNullNotAllowedFailExpression(column.getName(), tableType);
+                }
+                else {
+                    expression = new Cast(new NullLiteral(), toSqlType(column.getType()));
+                }
             }
             else {
                 Symbol input = visibleFieldMappings.get(index);
-                Type tableType = column.getType();
                 Type queryType = symbolAllocator.getTypes().get(input);
 
                 if (queryType.equals(tableType) || typeCoercion.isTypeOnlyCoercion(queryType, tableType)) {
-                    assignments.put(output, input.toSymbolReference());
+                    expression = input.toSymbolReference();
                 }
                 else {
-                    Expression cast = noTruncationCast(input.toSymbolReference(), queryType, tableType);
-                    assignments.put(output, cast);
+                    expression = noTruncationCast(input.toSymbolReference(), queryType, tableType);
                 }
-                insertedColumnsBuilder.add(column);
             }
+            if (!column.isNullable()) {
+                expression = new IfExpression(
+                        new IsNotNullPredicate(expression),
+                        expression,
```

Comment:
I am concerned this can result in `expression` being evaluated twice.
However, we can replace `IF(exp IS NOT NULL, exp, fail(...))` 
with `coalesce(exp, fail(...))`.
Per spec, `coalesce` should evaluate lazily.

Reply:
that's smart, thanks